File: internal/builtin/nodeutil/wasi.go
package nodeutil

// node:wasi 内置模块——实验性（stability 1）的 WASI preview1 方法面。
//
// aluka 无 WebAssembly 运行时（ADR: docs/adr/WASI-WASM.md），因此 WASI 只提供
// 与 Node 22 一致的类/方法面：
//   - WASI(options)：校验 options（version/args/env/preopens/stdin/stdout/
//     stderr），错误码/消息与 Node 对齐（ERR_INVALID_ARG_TYPE /
//     ERR_INVALID_ARG_VALUE / ERR_OUT_OF_RANGE）
//   - instance.wasiImport：46 个 preview1 系统调用函数（调用前抛
//     ERR_WASI_NOT_STARTED，Node 语义）
//   - start(instance) / initialize(instance)：校验顺序与 Node 一致
//     （started 标记 → instance/exports/memory 校验 → _start/_initialize
//     校验），内存校验后真正执行 WASM 需要运行时，属非目标
//   - getImportObject()：{ wasi_snapshot_preview1 | wasi_unstable: wasiImport }
//
// 已知差异（knownDifference）：
//   - 无 WASM 运行时：start/initialize 无法真正执行 WASM；wasiImport 的
//     46 个函数在 start 后调用会抛错（Node 会在真实实例上执行系统调用）。
//   - Node 在模块加载时输出 ExperimentalWarning；aluka 不输出（避免污染
//     差分输出），实验地位由 ADR 与本文件注释标记。
//   - unstable 版本的函数名以 preview1 集合近似（Node 内部同名绑定）。

import (
	"fmt"
	"math"

	"github.com/aluka-lang/aluka/internal/engine"
)

// wasiPreview1Functions 是 WASI preview1 导入对象的 46 个系统调用函数名。
var wasiPreview1Functions = []string{
	"args_get", "args_sizes_get", "clock_res_get", "clock_time_get",
	"environ_get", "environ_sizes_get", "fd_advise", "fd_allocate",
	"fd_close", "fd_datasync", "fd_fdstat_get", "fd_fdstat_set_flags",
	"fd_fdstat_set_rights", "fd_filestat_get", "fd_filestat_set_size",
	"fd_filestat_set_times", "fd_pread", "fd_prestat_dir_name", "fd_prestat_get",
	"fd_pwrite", "fd_read", "fd_readdir", "fd_renumber", "fd_seek", "fd_sync",
	"fd_tell", "fd_write", "path_create_directory", "path_filestat_get",
	"path_filestat_set_times", "path_link", "path_open", "path_readlink",
	"path_remove_directory", "path_rename", "path_symlink", "path_unlink_file",
	"poll_oneoff", "proc_exit", "proc_raise", "random_get", "sched_yield",
	"sock_accept", "sock_recv", "sock_send", "sock_shutdown",
}

// wasiCodeError 携带 Node 风格错误码（.code），name 按底层 error 决定
// （TypeError/RangeError/Error）。
type wasiCodeError struct {
	code string
	msg  string
	err  error
}

func (e *wasiCodeError) Error() string { return e.msg }
func (e *wasiCodeError) Code() string  { return e.code }
func (e *wasiCodeError) Unwrap() error { return e.err }

func wasiErrType(code, msg string) error {
	return &wasiCodeError{code: code, msg: msg, err: engine.ErrTypeError}
}

func wasiErrRange(code, msg string) error {
	return &wasiCodeError{code: code, msg: msg, err: engine.ErrRangeError}
}

func wasiErrPlain(code, msg string) error {
	return &wasiCodeError{code: code, msg: msg}
}

// wasiTypeString 渲染 Node 的 "Received ..." 片段。
func wasiTypeString(v engine.Value) string {
	switch v.Type() {
	case engine.TypeUndefined:
		return "undefined"
	case engine.TypeNull:
		return "null"
	case engine.TypeString:
		return "type string ('" + v.String() + "')"
	case engine.TypeNumber:
		return "type number (" + v.String() + ")"
	case engine.TypeBoolean:
		return "type boolean (" + v.String() + ")"
	case engine.TypeFunction:
		return "type function"
	default:
		return "type object"
	}
}

// wasiInstance 封装 WASI 实例状态。
type wasiInstance struct {
	started     bool
	bindingName string
	wasiImport  engine.Object
}

// NewWASI 构造 node:wasi 模块导出对象（{ WASI }）。
func NewWASI(ctx engine.Context) (engine.Value, error) {
	m := engine.NewObject()

	proto := engine.NewObject()
	ctor := engine.NewFunction("WASI", func(args []engine.Value) (engine.Value, error) {
		options := engine.Undefined()
		if len(args) > 0 {
			options = args[0]
		}
		inst, err := newWASIInstance(ctx, options)
		if err != nil {
			return engine.Undefined(), err
		}
		// 实例原型链接到 WASI.prototype（支持 instanceof）。
		engine.SetProto(inst, proto)
		return inst, nil
	})
	ctorObj, _ := ctor.AsObject()
	_ = proto.Set("constructor", ctor)
	_ = ctorObj.Set("prototype", proto)
	_ = m.Set("WASI", ctor)

	return m, nil
}

// newWASIInstance 构造一个 WASI 实例（构造函数主体）。
func newWASIInstance(ctx engine.Context, options engine.Value) (engine.Value, error) {
	opts := engine.NewObject()
	if !options.IsUndefined() && !options.IsNull() {
		o, ok := options.AsObject()
		if !ok {
			return engine.Undefined(), wasiErrType("ERR_INVALID_ARG_TYPE",
				fmt.Sprintf("The \"options\" argument must be of type object. Received %s", wasiTypeString(options)))
		}
		opts = o
	}

	// options.version：必填字符串。
	verV, _ := opts.Get("version")
	if verV.Type() != engine.TypeString {
		return engine.Undefined(), wasiErrType("ERR_INVALID_ARG_TYPE",
			fmt.Sprintf("The \"options.version\" property must be of type string. Received %s", wasiTypeString(verV)))
	}
	version := verV.String()
	var bindingName string
	switch version {
	case "unstable":
		bindingName = "wasi_unstable"
	case "preview1":
		bindingName = "wasi_snapshot_preview1"
	default:
		return engine.Undefined(), wasiErrType("ERR_INVALID_ARG_VALUE",
			fmt.Sprintf("The property 'options.version' unsupported WASI version. Received '%s'", version))
	}

	// options.args：数组。
	if av, err := opts.Get("args"); err == nil && !av.IsUndefined() {
		if _, ok := av.(*engine.ArrayValue); !ok {
			return engine.Undefined(), wasiErrType("ERR_INVALID_ARG_TYPE",
				fmt.Sprintf("The \"options.args\" property must be an instance of Array. Received %s", wasiTypeString(av)))
		}
	}
	// options.env / options.preopens：对象。
	for _, k := range []string{"env", "preopens"} {
		if ev, err := opts.Get(k); err == nil && !ev.IsUndefined() {
			if _, ok := ev.AsObject(); !ok {
				return engine.Undefined(), wasiErrType("ERR_INVALID_ARG_TYPE",
					fmt.Sprintf("The \"options.%s\" property must be of type object. Received %s", k, wasiTypeString(ev)))
			}
		}
	}
	// options.stdin/stdout/stderr：0..2^31-1 整数。
	for _, k := range []string{"stdin", "stdout", "stderr"} {
		v, _ := opts.Get(k)
		if v.IsUndefined() {
			continue
		}
		n, ok := v.Int()
		if !ok || n < 0 || n > math.MaxInt32 {
			received := n
			if !ok {
				received = -1
			}
			return engine.Undefined(), wasiErrRange("ERR_OUT_OF_RANGE",
				fmt.Sprintf("The value of \"options.%s\" is out of range. It must be >= 0 && <= 2147483647. Received %d", k, received))
		}
	}

	// wasiImport：46 个系统调用函数。Node 中 start() 成功（kInstance 设置）
	// 前调用一律抛 ERR_WASI_NOT_STARTED；aluka 无 WASM 运行时永远无法成功
	// start，故始终抛该错误（与 Node 可达状态一致）。
	wasiImport := engine.NewObject()
	for _, name := range wasiPreview1Functions {
		_ = wasiImport.Set(name, engine.NewFunction(name, func(a []engine.Value) (engine.Value, error) {
			return engine.Undefined(), wasiErrPlain("ERR_WASI_NOT_STARTED", "wasi.start() has not been called")
		}))
	}
	inst := &wasiInstance{
		started:     false,
		bindingName: bindingName,
		wasiImport:  wasiImport,
	}

	obj := engine.NewObject()
	_ = obj.Set("wasiImport", wasiImport)

	// start(instance)：started 标记 → instance/exports/memory → _start/_initialize。
	_ = obj.Set("start", engine.NewFunction("start", func(args []engine.Value) (engine.Value, error) {
		if inst.started {
			return engine.Undefined(), wasiErrPlain("ERR_WASI_ALREADY_STARTED", "WASI instance has already started")
		}
		inst.started = true
		if err := wasiSetupInstance(ctx, args); err != nil {
			return engine.Undefined(), err
		}
		// 无 WASM 运行时：即使内存校验通过也无法执行 _start。
		return engine.Undefined(), wasiErrPlain("ERR_WASI_NOT_IMPLEMENTED",
			"aluka: WASI _start requires a WebAssembly runtime (see docs/adr/WASI-WASM.md)")
	}))

	// initialize(instance)：started 标记 → instance/exports/memory → _initialize。
	_ = obj.Set("initialize", engine.NewFunction("initialize", func(args []engine.Value) (engine.Value, error) {
		if inst.started {
			return engine.Undefined(), wasiErrPlain("ERR_WASI_ALREADY_STARTED", "WASI instance has already started")
		}
		inst.started = true
		if err := wasiSetupInstance(ctx, args); err != nil {
			return engine.Undefined(), err
		}
		return engine.Undefined(), wasiErrPlain("ERR_WASI_NOT_IMPLEMENTED",
			"aluka: WASI _initialize requires a WebAssembly runtime (see docs/adr/WASI-WASM.md)")
	}))

	// getImportObject()：{ [bindingName]: wasiImport }。
	_ = obj.Set("getImportObject", engine.NewFunction("getImportObject", func(args []engine.Value) (engine.Value, error) {
		io := engine.NewObject()
		_ = io.Set(inst.bindingName, wasiImport)
		return io, nil
	}))

	return obj, nil
}

// wasiSetupInstance 执行 Node setupInstance 的校验：instance/exports 为对象、
// exports.memory 为 WebAssembly.Memory（aluka 无 Memory 类，任何内存值都
// 无法通过——以错误路径对齐 Node）。
func wasiSetupInstance(ctx engine.Context, args []engine.Value) error {
	instance := engine.Undefined()
	if len(args) > 0 {
		instance = args[0]
	}
	if _, ok := instance.AsObject(); !ok {
		return wasiErrType("ERR_INVALID_ARG_TYPE",
			fmt.Sprintf("The \"instance\" argument must be of type object. Received %s", wasiTypeString(instance)))
	}
	exports, _ := instance.AsObject()
	exportsVal, _ := exports.Get("exports")
	if _, ok := exportsVal.AsObject(); !ok {
		return wasiErrType("ERR_INVALID_ARG_TYPE",
			fmt.Sprintf("The \"instance.exports\" property must be of type object. Received %s", wasiTypeString(exportsVal)))
	}
	memory, _ := exportsVal.AsObject()
	_, _ = memory.Get("memory")
	// WebAssembly.Memory 类在 aluka 不存在 → 无法通过校验（Node 需要
	// instanceof WebAssembly.Memory）。
	return wasiErrType("ERR_INVALID_ARG_TYPE",
		"\"instance.exports.memory\" property must be a WebAssembly.Memory object")
}

File: aluka_r/crates/aluka-vm/src/builtins/wasi.rs
//! `wasi` 内置模块（Phase 8）：实验性（stability 1）的 WASI preview1 方法面。
//!
//! 逐函数移植 Go oracle（`nodeutil/wasi.go`）。aluka 无 WebAssembly 运行时，
//! 因此 WASI 只提供与 Node 22 一致的类/方法面：
//! - `WASI(options)`：校验 options（version/args/env/preopens/stdin/stdout/
//!   stderr），错误码/消息与 Node 对齐（ERR_INVALID_ARG_TYPE /
//!   ERR_INVALID_ARG_VALUE / ERR_OUT_OF_RANGE）；
//! - `instance.wasiImport`：46 个 preview1 系统调用函数（调用前抛
//!   ERR_WASI_NOT_STARTED，Node 语义）；
//! - `start(instance)` / `initialize(instance)`：校验顺序与 Node 一致
//!   （started 标记 → instance/exports/memory 校验 → _start/_initialize）；
//! - `getImportObject()`：`{ wasi_snapshot_preview1 | wasi_unstable: wasiImport }`。
//!
//! 已知差异（对齐 Go 实现自身的 knownDifference）：无 WASM 运行时，
//! `start`/`initialize` 在内存校验后无法真正执行 WASM，抛
//! ERR_WASI_NOT_IMPLEMENTED；实验警告不输出（避免污染差分输出）。

use crate::builtins::{
    BuiltinRegistry, ModuleDef, current_receiver, register_handler, set_module_prop,
};
use crate::heap::HeapObject;
use crate::interpreter::{Vm, VmError};
use crate::value::Value;
use aluka_core::ObjectRef;

/// `require("wasi")` / `require("node:wasi")` 模块条目。
pub const MODULE: ModuleDef = ModuleDef {
    name: "wasi",
    build,
};

/// WASI preview1 导入对象的 46 个系统调用函数名。
const PREVIEW1_FUNCTIONS: [&str; 46] = [
    "args_get",
    "args_sizes_get",
    "clock_res_get",
    "clock_time_get",
    "environ_get",
    "environ_sizes_get",
    "fd_advise",
    "fd_allocate",
    "fd_close",
    "fd_datasync",
    "fd_fdstat_get",
    "fd_fdstat_set_flags",
    "fd_fdstat_set_rights",
    "fd_filestat_get",
    "fd_filestat_set_size",
    "fd_filestat_set_times",
    "fd_pread",
    "fd_prestat_dir_name",
    "fd_prestat_get",
    "fd_pwrite",
    "fd_read",
    "fd_readdir",
    "fd_renumber",
    "fd_seek",
    "fd_sync",
    "fd_tell",
    "fd_write",
    "path_create_directory",
    "path_filestat_get",
    "path_filestat_set_times",
    "path_link",
    "path_open",
    "path_readlink",
    "path_remove_directory",
    "path_rename",
    "path_symlink",
    "path_unlink_file",
    "poll_oneoff",
    "proc_exit",
    "proc_raise",
    "random_get",
    "sched_yield",
    "sock_accept",
    "sock_recv",
    "sock_send",
    "sock_shutdown",
];

/// 构建 node:wasi 模块导出对象（`{ WASI }`）。
fn build(vm: &mut Vm, registry: &mut BuiltinRegistry) -> Result<ObjectRef, VmError> {
    let obj = vm.alloc_ordinary();

    // WASI 类：NativeCtor 名称即分派键——`new WASI(opts)` 经 do_construct
    // 的 `lookup(ctor_name)` 命中处理器；prototype 挂 constructor 支持 instanceof。
    let proto = vm.alloc_ordinary();
    let ctor = vm.alloc_native_ctor("wasi.WASI.ctor", None);
    let ctor_val = Value::Object(ctor);
    let proto_val = Value::Object(proto);
    let _ = vm.set_property(Value::Object(proto), "constructor", ctor_val);
    let _ = vm.set_property(Value::Object(ctor), "prototype", proto_val);
    let _ = vm.set_property(Value::Object(obj), "WASI", ctor_val);

    register_handler(registry, "wasi.WASI", "ctor", wasi_ctor);
    register_handler(registry, "wasi:inst", "start", wasi_start);
    register_handler(registry, "wasi:inst", "initialize", wasi_initialize);
    register_handler(
        registry,
        "wasi:inst",
        "getImportObject",
        wasi_get_import_object,
    );
    for name in PREVIEW1_FUNCTIONS {
        register_handler(registry, "wasi:import", name, wasi_import_stub);
    }

    Ok(obj)
}

/// 构造携带 Node 风格错误码（.code）的异常对象。
fn code_error(vm: &mut Vm, code: &str, msg: &str) -> VmError {
    let err = vm.alloc_ordinary();
    let msg_val = Value::Object(vm.alloc_string(msg.to_owned()));
    let _ = vm.set_property(Value::Object(err), "message", msg_val);
    let code_val = Value::Object(vm.alloc_string(code.to_owned()));
    let _ = vm.set_property(Value::Object(err), "code", code_val);
    VmError::Thrown(Value::Object(err))
}

/// 是否为堆字符串值。
fn is_string(vm: &Vm, v: Value) -> bool {
    matches!(v, Value::Object(r)
        if matches!(vm.heap.get(r.index()), Some(HeapObject::String(_))))
}

/// 是否为普通对象（Ordinary 堆对象）。
fn is_ordinary(vm: &Vm, v: Value) -> bool {
    matches!(v, Value::Object(r)
        if matches!(vm.heap.get(r.index()), Some(HeapObject::Ordinary { .. })))
}

/// 渲染 Node 的 "Received ..." 片段（对齐 Go `wasiTypeString`）。
fn type_string(vm: &Vm, v: Value) -> String {
    match v {
        Value::Undefined => "undefined".to_owned(),
        Value::Null => "null".to_owned(),
        Value::Number(n) => format!("type number ({})", vm.format_value(Value::Number(n))),
        Value::Boolean(b) => format!("type boolean ({b})"),
        Value::Object(r) => match vm.heap.get(r.index()) {
            Some(HeapObject::String(s)) => format!("type string ('{s}')"),
            Some(HeapObject::Closure { .. })
            | Some(HeapObject::NativeCtor { .. })
            | Some(HeapObject::NativeFn { .. }) => "type function".to_owned(),
            _ => "type object".to_owned(),
        },
    }
}

/// `new WASI(options)`：构造函数主体（校验 + 实例装配）。
fn wasi_ctor(vm: &mut Vm, args: &[Value]) -> Result<Value, VmError> {
    let options = args.first().copied().unwrap_or(Value::Undefined);
    let opts_val = match options {
        Value::Undefined | Value::Null => None,
        v if is_ordinary(vm, v) => Some(v),
        _ => {
            return Err(code_error(
                vm,
                "ERR_INVALID_ARG_TYPE",
                &format!(
                    "The \"options\" argument must be of type object. Received {}",
                    type_string(vm, options)
                ),
            ));
        }
    };

    // options.version：必填字符串。
    let version = match opts_val {
        Some(o) => {
            let ver = vm.get_property(o, "version")?;
            if is_string(vm, ver) {
                let Value::Object(s) = ver else {
                    unreachable!()
                };
                match vm.heap.get(s.index()) {
                    Some(HeapObject::String(text)) => text.clone(),
                    _ => String::new(),
                }
            } else {
                return Err(code_error(
                    vm,
                    "ERR_INVALID_ARG_TYPE",
                    &format!(
                        "The \"options.version\" property must be of type string. Received {}",
                        type_string(vm, ver)
                    ),
                ));
            }
        }
        None => {
            return Err(code_error(
                vm,
                "ERR_INVALID_ARG_TYPE",
                "The \"options.version\" property must be of type string. Received undefined",
            ));
        }
    };
    let binding_name = match version.as_str() {
        "unstable" => "wasi_unstable",
        "preview1" => "wasi_snapshot_preview1",
        other => {
            return Err(code_error(
                vm,
                "ERR_INVALID_ARG_VALUE",
                &format!(
                    "The property 'options.version' unsupported WASI version. Received '{other}'"
                ),
            ));
        }
    };

    // options.args：数组。
    if let Some(o) = opts_val {
        let av = vm.get_property(o, "args")?;
        if !matches!(av, Value::Undefined) {
            let is_array = matches!(av, Value::Object(ar)
                if matches!(vm.heap.get(ar.index()), Some(HeapObject::Array { .. })));
            if !is_array {
                return Err(code_error(
                    vm,
                    "ERR_INVALID_ARG_TYPE",
                    &format!(
                        "The \"options.args\" property must be an instance of Array. Received {}",
                        type_string(vm, av)
                    ),
                ));
            }
        }
    }

    // options.env / options.preopens：对象。
    for key in ["env", "preopens"] {
        if let Some(o) = opts_val {
            let ev = vm.get_property(o, key)?;
            if !matches!(ev, Value::Undefined) && !is_ordinary(vm, ev) {
                return Err(code_error(
                    vm,
                    "ERR_INVALID_ARG_TYPE",
                    &format!(
                        "The \"options.{key}\" property must be of type object. Received {}",
                        type_string(vm, ev)
                    ),
                ));
            }
        }
    }

    // options.stdin/stdout/stderr：0..2^31-1 整数。
    for key in ["stdin", "stdout", "stderr"] {
        let v = match opts_val {
            Some(o) => vm.get_property(o, key)?,
            None => Value::Undefined,
        };
        if matches!(v, Value::Undefined) {
            continue;
        }
        let ok_int = matches!(v, Value::Number(n)
            if n.fract() == 0.0 && (0.0..=2147483647.0).contains(&n));
        if !ok_int {
            let received = match v {
                Value::Number(n) => n as i64,
                _ => -1,
            };
            return Err(code_error(
                vm,
                "ERR_OUT_OF_RANGE",
                &format!(
                    "The value of \"options.{key}\" is out of range. It must be >= 0 && <= 2147483647. Received {received}"
                ),
            ));
        }
    }

    // 实例对象：原型链接到 WASI.prototype（支持 instanceof）；
    // `_builtinNs` 实例分派（引擎 CALL_METHOD 无属性回退）。
    let proto_ref = wasi_proto(vm);
    let inst = vm.alloc_ordinary_with_proto(proto_ref);
    let ns_val = Value::Object(vm.alloc_string("wasi:inst".to_owned()));
    set_module_prop(vm, inst, "_builtinNs", ns_val)?;
    set_module_prop(vm, inst, "_started", Value::Boolean(false))?;
    let binding_val = Value::Object(vm.alloc_string(binding_name.to_owned()));
    set_module_prop(vm, inst, "_bindingName", binding_val)?;

    // wasiImport：46 个系统调用函数（每实例新建，对齐 Go）。
    let wasi_import = vm.alloc_ordinary();
    let import_ns = Value::Object(vm.alloc_string("wasi:import".to_owned()));
    let _ = vm.set_property(Value::Object(wasi_import), "_builtinNs", import_ns);
    for name in PREVIEW1_FUNCTIONS {
        let fn_ref = vm.alloc_native_fn(&format!("wasi:import.{name}"));
        set_module_prop(vm, wasi_import, name, Value::Object(fn_ref))?;
    }
    set_module_prop(vm, inst, "wasiImport", Value::Object(wasi_import))?;

    for (prop, name) in [
        ("start", "wasi:inst.start"),
        ("initialize", "wasi:inst.initialize"),
        ("getImportObject", "wasi:inst.getImportObject"),
    ] {
        let fn_ref = vm.alloc_native_fn(name);
        set_module_prop(vm, inst, prop, Value::Object(fn_ref))?;
    }

    Ok(Value::Object(inst))
}

/// 读取模块单例上 WASI 类的 prototype 对象。
fn wasi_proto(vm: &mut Vm) -> Option<ObjectRef> {
    let module = vm.builtin_registry.module("wasi")?;
    let ctor = vm.get_property(Value::Object(module), "WASI").ok()?;
    let proto = vm.get_property(ctor, "prototype").ok()?;
    match proto {
        Value::Object(p) => Some(p),
        _ => None,
    }
}

/// `start(instance)` / `initialize(instance)` 共用的 setupInstance 校验：
/// instance/exports 为对象、exports.memory 为 WebAssembly.Memory（aluka 无
/// Memory 类，任何内存值都无法通过——以错误路径对齐 Node）。
fn setup_instance(vm: &mut Vm, args: &[Value]) -> Result<(), VmError> {
    let instance = args.first().copied().unwrap_or(Value::Undefined);
    if !is_ordinary(vm, instance) {
        return Err(code_error(
            vm,
            "ERR_INVALID_ARG_TYPE",
            &format!(
                "The \"instance\" argument must be of type object. Received {}",
                type_string(vm, instance)
            ),
        ));
    }
    let exports = vm.get_property(instance, "exports")?;
    if !is_ordinary(vm, exports) {
        return Err(code_error(
            vm,
            "ERR_INVALID_ARG_TYPE",
            &format!(
                "The \"instance.exports\" property must be of type object. Received {}",
                type_string(vm, exports)
            ),
        ));
    }
    let _ = vm.get_property(exports, "memory")?;
    Err(code_error(
        vm,
        "ERR_INVALID_ARG_TYPE",
        "\"instance.exports.memory\" property must be a WebAssembly.Memory object",
    ))
}

/// started 标记检查 + 置位（对齐 Go：先置位后校验，失败路径保持已启动）。
fn check_and_mark_started(vm: &mut Vm) -> Result<(), VmError> {
    let receiver = current_receiver();
    if matches!(vm.get_property(receiver, "_started")?, Value::Boolean(true)) {
        return Err(code_error(
            vm,
            "ERR_WASI_ALREADY_STARTED",
            "WASI instance has already started",
        ));
    }
    let _ = vm.set_property(receiver, "_started", Value::Boolean(true));
    Ok(())
}

/// `wasi.start(instance)`。
fn wasi_start(vm: &mut Vm, args: &[Value]) -> Result<Value, VmError> {
    check_and_mark_started(vm)?;
    setup_instance(vm, args)?;
    // 无 WASM 运行时：即使内存校验通过也无法执行 _start。
    Err(code_error(
        vm,
        "ERR_WASI_NOT_IMPLEMENTED",
        "aluka: WASI _start requires a WebAssembly runtime (see docs/adr/WASI-WASM.md)",
    ))
}

/// `wasi.initialize(instance)`。
fn wasi_initialize(vm: &mut Vm, args: &[Value]) -> Result<Value, VmError> {
    check_and_mark_started(vm)?;
    setup_instance(vm, args)?;
    Err(code_error(
        vm,
        "ERR_WASI_NOT_IMPLEMENTED",
        "aluka: WASI _initialize requires a WebAssembly runtime (see docs/adr/WASI-WASM.md)",
    ))
}

/// `wasi.getImportObject()`：`{ [bindingName]: wasiImport }`。
fn wasi_get_import_object(vm: &mut Vm, _args: &[Value]) -> Result<Value, VmError> {
    let receiver = current_receiver();
    let binding = vm.get_property(receiver, "_bindingName")?;
    let wasi_import = vm.get_property(receiver, "wasiImport")?;
    let io = vm.alloc_ordinary();
    let _ = vm.set_property(Value::Object(io), &vm.format_value(binding), wasi_import);
    Ok(Value::Object(io))
}

/// wasiImport 46 个系统调用 stub：始终抛 ERR_WASI_NOT_STARTED。
fn wasi_import_stub(vm: &mut Vm, _args: &[Value]) -> Result<Value, VmError> {
    Err(code_error(
        vm,
        "ERR_WASI_NOT_STARTED",
        "wasi.start() has not been called",
    ))
}
